aten/src/ATen/cuda/PeerToPeerAccess.cpp

```diff
@@ -49,13 +55,32 @@ bool get_p2p_access(int dev, int dev_to_access) {
   int access = 0;
   C10_CUDA_CHECK(cudaDeviceCanAccessPeer(&access, dev, dev_to_access));
   if (access) {
-    cudaError_t err = cudaDeviceEnablePeerAccess(dev_to_access, 0);
-    if (err == cudaErrorPeerAccessAlreadyEnabled) {
-      // ignore and clear the error if access was already enabled
-      cudaGetLastError();
+#if CUDA_VERSION >= 11040
+    if (using_cudaMallocAsync) {
+      // cudaMallocAsync pools are unaffected by cudaDeviceEnablePeerAccess.
+      // We need pool-specific enablement. See
+      // https://developer.nvidia.com/blog/using-cuda-stream-ordered-memory-allocator-part-2/
+      cudaMemPool_t mempool;
+      C10_CUDA_CHECK(cudaDeviceGetDefaultMemPool(&mempool, dev_to_access));
+      cudaMemAccessDesc desc = {};
+      desc.location.type = cudaMemLocationTypeDevice;
+      desc.location.id = dev;
+      desc.flags = cudaMemAccessFlagsProtReadWrite;
+      C10_CUDA_CHECK(cudaMemPoolSetAccess(mempool, &desc, 1 /* numDescs */));
     } else {
-      C10_CUDA_CHECK(err);
+      TORCH_INTERNAL_ASSERT(std::strcmp(c10::cuda::CUDACachingAllocator::allocatorBackend(),
+                                        "native") == 0);
+#endif
+      cudaError_t err = cudaDeviceEnablePeerAccess(dev_to_access, 0);
+      if (err == cudaErrorPeerAccessAlreadyEnabled) {
+        // ignore and clear the error if access was already enabled
+        cudaGetLastError();
+      } else {
+        C10_CUDA_CHECK(err);
+      }
+#if CUDA_VERSION >= 11040
     }
+#endif
```

Comment:
Stubbing out a single curly here is pretty hard to understand. Two recommendations:

1. After you fix CUDACachingAllocator::allocatorBackend() to not take a string, also make it return false if CUDA version isn't high enough (and I'd recommend TORCH_WARN_ONCE that user attempted to use cudaMallocAsync but it doesn't work). Actually, I might even suggest having it error in that case (you asked for this allocator but we don't have it, crash)
2. If everything in the body of the if below can compile, unconditionally compile it. If it doesn't, if around the inside of the if statement only (and when it's not there TORCH_INTERNAL_ASSERT(false). It is slightly slower than what you have here but it doesn't matter once everyone upgrades CUDA enough

Reply:
Done, I think. Please double check if I got the right idea.